File: src/generateMetric.php
<?php  
# loading the thread sizes and pitches
$threads = json_decode(file_get_contents('threads.json'),true);

# the thread angles    overhang angle = 90 - angle/2
$angles = [90,80,70,60,50]; 

# the maximum +- tolerance
$tolMax = 0.5;

# The tolerance step size 
$tolStep = 0.025;

foreach ($angles as $angle) {
	$filename = '../FDM' . $angle . 'MetricTrapezoidalThreads.xml';
	$name = 'FDM ' . $angle . ' Degree Metric Trapezoidal Threads';

	$threadType = new SimpleXMLElement('<?xml version="1.0" encoding="UTF-8"?><ThreadType></ThreadType>');
	$threadType->addChild('Name', $name); 
	$threadType->addChild('CustomName', $name); 
	$threadType->addChild('Unit', 'mm'); 
	$threadType->addChild('Angle', $angle);
	$threadType->addChild('SortOrder', 4); 

	foreach ($threads as $size => $pitches) {
		$threadSize = $threadType->addChild('ThreadSize'); 
		$threadSize->addChild('Size', $size);

		foreach ($pitches as $pitch) {
			$designation = $threadSize->addChild('Designation');
			$designation->addChild('ThreadDesignation', 'FDM' . $angle . '-' . $size . 'x' . $pitch);
			$designation->addChild('CTD', 'FDM' . $angle . '-' . $size . 'x' . $pitch);
			$designation->addChild('Pitch', $pitch);

			/*
			 * $pitch = thread pitch
			 * $angle = thread form angle
			 * $size = nominal diameter of the thread
			 * $radius = nominal radius of the thread
			 * $height = nominal height of the thread form
			 * $crestH = distance between the crest and the top of the v thread form
			 * $pitchH = distance between the pitch radius and the top of the v thread form
			 * $rootH = distance between the root and the bottom of the v thread form
			 * $MajorRadius = the major radius of the nominal thread
			 * $pitchRadius = the pitch radius of the nominal thread
			 * $minorRadius = the minor radius of the nominal thread
			 */
			$radius = $size/2;
			$height = tan(deg2rad(90-($angle/2)))*($pitch/2);

			$crestH = tan(deg2rad(90-($angle/2)))*($pitch/8);
			$pitchH = tan(deg2rad(90-($angle/2)))*($pitch/4);
			$rootH = tan(deg2rad(90-($angle/2)))*($pitch/8);

			$MajorRadius = $radius;
			$pitchRadius = $radius-$pitchH+$crestH;
			$minorRadius = $radius-$height+$rootH+$crestH;

			$tol = 0;
			while ( $tol <= $tolMax && $tol <= $crestH && $tol <= $rootH) {
				/*
				 * $tol = the tolerance between the actual thread and the nominal thread
				 * $externalMajorD = the major diameter of the external thread
				 * $externalPitchD = the pitch diameter of the external thread
				 * $externalMinorD = the minor diameter of the external thread
				 * $internalMajorD = the major diameter of the internal thread
				 * $internalPitchD = the pitch diameter of the internal thread
				 * $internalMinorD = the minor diameter of the internal thread
				 */
				$externalMajorD = ($MajorRadius-$tol)*2;
				$externalPitchD = ($pitchRadius-($tol/sin(deg2rad($angle/2))))*2;
				$externalMinorD = ($minorRadius-$tol)*2;
				$internalMajorD = ($MajorRadius+$tol)*2;
				$internalPitchD = ($pitchRadius+($tol/sin(deg2rad($angle/2))))*2;
				$internalMinorD = ($minorRadius+$tol)*2;

				$thread = $designation->addChild('Thread');
				$thread->addChild('Gender', 'external');
				$thread->addChild('Class', $tol . 'e');
				$thread->addChild('MajorDia', $externalMajorD);
				$thread->addChild('PitchDia', $externalPitchD);
				$thread->addChild('MinorDia', $externalMinorD);

				$thread = $designation->addChild('Thread');
				$thread->addChild('Gender', 'internal');
				$thread->addChild('Class', $tol . 'i');
				$thread->addChild('MajorDia', $internalMajorD);
				$thread->addChild('PitchDia', $internalPitchD);
				$thread->addChild('MinorDia', $internalMinorD);

				$tol += $tolStep;
			}
		}
	}

	$dom = new DOMDocument('1.0');
	$dom->preserveWhiteSpace = false;
	$dom->formatOutput = true;
	$dom->loadXML($threadType->asXML());
	$dom->save($filename);
}
?>

File: src/generateMetric.py
import json
import math
import xml.etree.ElementTree as ET

# Loading the thread sizes and pitches
with open('threads.json', 'r') as f:
    threads = json.load(f)

# The thread angles overhang angle = 90 - angle/2
angles = [90, 80, 70, 60, 50]

# The maximum +- tolerance
tolMax = 0.5

# The tolerance step size
tolStep = 0.025

for angle in angles:
    filename = f'../FDM{angle}MetricTrapezoidalThreads.xml'
    name = f'FDM {angle} Degree Metric Trapezoidal Threads'

    # Creating the root XML element
    threadType = ET.Element('ThreadType')
    ET.SubElement(threadType, 'Name').text = name
    ET.SubElement(threadType, 'CustomName').text = name
    ET.SubElement(threadType, 'Unit').text = 'mm'
    ET.SubElement(threadType, 'Angle').text = str(angle)
    ET.SubElement(threadType, 'SortOrder').text = '4'

    for size, pitches in threads.items():
        threadSize = ET.SubElement(threadType, 'ThreadSize')
        ET.SubElement(threadSize, 'Size').text = f'{float(size):.1f}'

        for pitch in pitches:
            designation = ET.SubElement(threadSize, 'Designation')
            designation_text = f'FDM{angle}-{size}x{pitch}'
            ET.SubElement(designation, 'ThreadDesignation').text = designation_text
            ET.SubElement(designation, 'CTD').text = designation_text
            ET.SubElement(designation, 'Pitch').text = f'{float(pitch):.1f}'

            radius = float(size) / 2
            height = math.tan(math.radians(90 - (angle / 2))) * (float(pitch) / 2)

            crestH = math.tan(math.radians(90 - (angle / 2))) * (float(pitch) / 8)
            pitchH = math.tan(math.radians(90 - (angle / 2))) * (float(pitch) / 4)
            rootH = math.tan(math.radians(90 - (angle / 2))) * (float(pitch) / 8)

            MajorRadius = radius
            pitchRadius = radius - pitchH + crestH
            minorRadius = radius - height + rootH + crestH

            tol = 0
            while (round(tol, 6) <= round(tolMax, 6) and
                   round(tol, 6) <= round(crestH, 6) and
                   round(tol, 6) <= round(rootH, 6)):
                externalMajorD = (MajorRadius - tol) * 2
                externalPitchD = (pitchRadius - (tol / math.sin(math.radians(angle / 2)))) * 2
                externalMinorD = (minorRadius - tol) * 2
                internalMajorD = (MajorRadius + tol) * 2
                internalPitchD = (pitchRadius + (tol / math.sin(math.radians(angle / 2)))) * 2
                internalMinorD = (minorRadius + tol) * 2

                thread = ET.SubElement(designation, 'Thread')
                ET.SubElement(thread, 'Gender').text = 'external'
                ET.SubElement(thread, 'Class').text = f'{tol:.3f}e'
                ET.SubElement(thread, 'MajorDia').text = f'{externalMajorD:.6f}'
                ET.SubElement(thread, 'PitchDia').text = f'{externalPitchD:.6f}'
                ET.SubElement(thread, 'MinorDia').text = f'{externalMinorD:.6f}'

                thread = ET.SubElement(designation, 'Thread')
                ET.SubElement(thread, 'Gender').text = 'internal'
                ET.SubElement(thread, 'Class').text = f'{tol:.3f}i'
                ET.SubElement(thread, 'MajorDia').text = f'{internalMajorD:.6f}'
                ET.SubElement(thread, 'PitchDia').text = f'{internalPitchD:.6f}'
                ET.SubElement(thread, 'MinorDia').text = f'{internalMinorD:.6f}'

                tol += tolStep

    # Saving the XML to file
    tree = ET.ElementTree(threadType)
    tree.write(filename, encoding='utf-8', xml_declaration=True)